### ISSUE TITLE
Compare both version side by side with same initial state and random seeds

### ISSUE THREAD

**Issue by rickstaa** [rickstaa/LAC-TF2-TORCH-translation#4]
# Validate learning function
:warning: THIS REPORT IS NO LONGER VALID DUE TO BUG FIX IN TARGET NETWORK INITIATION  c7a83d267a58df468d3c0240f3286cd2462fa975 :warning:

Below you will find a validation for the losses, and STG steps between the Tensorflow and Pytorch version. I did this while setting the initial state, random seeds, weights and biases equal in both version.

### Initial states and states from the environment

:heavy_check_mark: Are equal.

### Initial network weights

:heavy_check_mark: Is equal

### First actions from the policy

:heavy_check_mark: Is equal

### Replay buffer before the first epoch

:x: Is slightly different? Same range.

This is caused because after the 2n episode, the action from the policy is different.

```python
a: [ 0.806744 -1.      ]
action: [  8.06744 -10.     ]
s: [ 1.28136525 -0.42225537]
s_: [ 1.07099065 -1.07482597]
terminal: 0.0
r: 0.05171322309503261
```

vs

```python
a: [-0.99590796 -1.        ]
action: [ -9.95908 -10.     ]
s: [ 1.28136525 -0.42225537]
s_: [ 1.44671609 -1.18728409]
terminal: 0.0
r: 0.3812822278448363
```

#### Debug

Weights are still the same and not updated. Meaning this is caused by the forward step used to calculate the current best action. May be caused by the different random seed between PyTorch and TensorFlow sample that is taken in the reparameterisation trick.

```python
base_distribution = tfp.distributions.MultivariateNormalDiag(loc=tf.zeros(self.a_dim), scale_diag=tf.ones(self.a_dim))
epsilon = base_distribution.sample(batch_size)
```

##### Solution

For a real comparison, I added the following code during the comparison.

```python
if USE_FIXED_SEED:
    a = torch.FloatTensor(1, 2).uniform_(-1, 1).squeeze().numpy()
else:
    a = policy.choose_action(s)
```

Now we get the same environment initialisation, replay batch, weights etc.

### L

:heavy_check_mark: Same shape and equal!

### L_

:heavy_check_mark: Same shape and equal!

### L_derta:

|    | tf        | PyT             |
|----|-----------|-----------------|
| 1  | 7249.67 | 7167.8168945312 |
| 2 | 9131.958            | 9283.3798828125                |
| 10 | 8690.137          | 8478.4843750000                |

### labda_loss

:heavy_check_mark: 

|    | tf        | PyT             |
|----|-----------|-----------------|
| 1  | 72.13658 | 72.0389022827 |
| 2 | 89.95388  | 92.3726654053                |
| 10 |  76.911026         | 76.7173843384                | 

### Labda
:heavy_check_mark: 
|    | tf        | PyT             |
|----|-----------|-----------------|
| 0  | 0.99 |  0.99 |
| 1  | 0.990099 |  0.9900990129 |
| 2 |  0.9901979  |  0.9901978970,               |
| 10 |   0.99118865       |   0.9910914302

### Alpha loss
:heavy_check_mark: 
|    | tf        | PyT             |
|----|-----------|-----------------|
| 1  | 2.6537108 | 2.6728796959, |
| 2 |  2.6005726           | 2.6294424534                 |
| 10 |  2.3576145         |  2.3356657028,              | 

This is expected since log_alpha start out as zero. Why not the cause in tf?

### log_alpha
:heavy_check_mark: 
|    | tf        | PyT             |
|----|-----------|-----------------|
| 0 | -0.01 | -0.01
| 1  | -0.009950325 | -0.0099503249
| 2 | -0.009850339            | -0.0098503400                |
| 10 | -0.0088536525          | -0.0089539969                | 

### alpha
:heavy_check_mark: 
|    | tf        | PyT             |
|----|-----------|-----------------|
| 0 | 0.99 | 0.99
| 1  | 0.990099 | 1.0
| 2 | 0.990198            | 0.9901980162                |
| 10 | 0.9911854           | 0.9910859466                 |

:warning: **Note!** self.alpha is updated, but alpha is not! Why is there an alpha variable that is always `1.0`?

:exclamation: Due to this line of code! :exclamation: 

```python
from variant import * 
```

It now imports all the variables in the variant file as globals. Now we dodged the bullet, but this causes confusion and can cause bad things to happen! Replace with:

```python
from variant_translated import VARIANT
```

### Actor loss

:heavy_check_mark: 
|    | tf        | PyT             |
|----|-----------|-----------------|
| 1  | 7443.927 | 7362.1450195312 |
| 2 |  9305.847           |  9456.0302734375               |
| 10 |  8879.486         | 8660.6162109375                |

### L_target

:heavy_check_mark: Same shape and equal!

### L_error

|    | tf        | PyT             |
|----|-----------|-----------------|
| 1  | 1096044800.0 | 1.1042696960e+09 |
| 25 | 411770270.0          |  4.1684979200e+08               |
| 10 |  306597000.0         |  304697820.0                |

### Entropy

|    | tf        | PyT             |
|----|-----------|-----------------|
| 1  | -302.3425 | -303.0217590332 |
| 25 | -266.00842          | -266.2569580078                |
| 10 |  -266.85327         |  -267.50128               |

### Ga

Updated: :+1: 
:heavy_check_mark: All layers have very similar (up till low-level digits) weights and biases. I checked that this was the case up till the 10th epoch.

### Lc

Updated: :+1:
:heavy_check_mark:  All layers have very similar (up till low-level digits) weights and biases. I checked that this was the case up till the 10th epoch.

### Ga_

:heavy_check_mark: Updated according to moving average. Also not equal to ga. I checked that this was the case up till the 10th epoch.

```python
(self.lc_copy.l_net[0].bias == self.lc_.l_net[0].bias).numpy().all()
False
```

### Lc_

:heavy_check_mark:  Updated according to moving average. Also not equal to Lc_. I checked that this was the case up till the 10th epoch.

```python
(self.lc_copy.l_net[0].bias == self.lc_.l_net[0].bias).numpy().all()
False
```

### Lya_ga_
 
:heavy_check_mark: 

This network is frozen does not update. I checked that this was the case up till the 10th epoch.

### lya_lc_

:heavy_check_mark:

This network is frozen does not update. I checked that this was the case up till the 10th epoch.

## Other problems!

When you set lambda or alpha to 1.0 the log_lambda becomes 0.0. This means the initial lambda and alpha losses should become 0.0 as well and the parameters not updated. Something strange happens in TensorFlow compared to PyTorch.

### Pytorch

![image](https://user-images.githubusercontent.com/17570430/92042449-86b4a480-ed7a-11ea-9d4e-ed0528561318.png)

There is however still a gradient, so the lambda gets updated:

![image](https://user-images.githubusercontent.com/17570430/92043260-0727d500-ed7c-11ea-9cf0-4d659b936847.png)

This might be due to how Adam deals with 0.0 losses and prevents an algorithm from being stuck when initialised with a parameter of 0.0 by adding some noise? I think this, therefore, should not cause a problem. If this behaviour however is not desired we can use:

```python
torch.mean(self.labda * self.l_derta.detach())
torch.mean(self.labda * self.l_derta.detach())
```

instead of:

```python
-torch.mean(self.log_labda * self.l_derta.detach())
torch.mean(self.alpha * (log_pis + self.target_entropy).detach())
```

### Tensorflow

![image](https://user-images.githubusercontent.com/17570430/92042501-a3e97300-ed7a-11ea-82bb-f81fb952fba6.png)

This is strange as log_labda also is 0.0. I have not yet found why this works in TensorFlow.
